tests/cases/colorLegend.js

```diff
@@ -0,0 +1,187 @@
+    expect($(legends[0]).find('svg g.tick').length).toBe(4);
+
+    expect($(legends[1]).find('svg>rect').length).toBe(9);
+    expect($(legends[1]).find('svg g.tick').length).toBe(5);
+
+    expect($(legends[2]).find('svg>rect').length).toBe(11);
+    expect($(legends[2]).find('svg g.tick').length).toBe(6);
+    expect($(legends[2]).find('svg g.tick:last text').text()).toBe('830k');
+
+    expect($(legends[3]).find('svg>rect').length).toBe(8);
+    expect($(legends[3]).find('svg g.tick').length).toBe(5);
+    expect($(legends[3]).find('svg g.tick:eq(2) text').text()).toBe('150');
+
+    expect($(legends[4]).find('svg>rect').length).toBe(8);
+    expect($(legends[4]).find('svg>rect:first').attr('fill')).toBe(allCategories[4].colors[0]);
+    expect($(legends[4]).find('svg g.tick:last text').text()).toBe('0.0');
+
+    expect($(legends[5]).find('svg>rect').length).toBe(1);
+    expect($(legends[5]).find('svg>rect:first').attr('fill').indexOf('url')).not.toBe(-1);
+    expect($(legends[5]).find('svg g.tick').length).toBe(6);
+    expect($(legends[5]).find('svg g.tick:first text').text()).toBe('0.00k');
+
+    expect($(legends[6]).find('svg>rect').length).toBe(1);
+    expect($(legends[6]).find('svg g.tick').length).toBe(5);
+
+    expect($(legends[7]).find('svg>rect').length).toBe(1);
+    expect($(legends[7]).find('svg g.tick').length).toBe(5);
+    expect($(legends[7]).find('svg g.tick text').toArray().map(function (text) {
+      return $(text).text();
+    }).join(', ')).toBe('150, 400, 1.1k, 3.0k, 8.1k');
+    done();
+  });
+
+  it('test mouse events', function (done) {
+    function CreateEvent(eventType, params) {
+      params = params || { bubbles: false, cancelable: false };
+      var mouseEvent = document.createEvent('MouseEvent');
+      mouseEvent.initMouseEvent(eventType, params.bubbles, params.cancelable, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
+      return mouseEvent;
+    }
+
+    Math.trunc = Math.trunc || function (x) {
+      if (isNaN(x)) {
+        return NaN;
+      }
+      if (x > 0) {
+        return Math.floor(x);
+      }
+      return Math.ceil(x);
+    };
+
+    container[0].dispatchEvent(CreateEvent('mouseenter'));
+    container[0].dispatchEvent(CreateEvent('mouseleave'));
+
+    legendWidget.categories([allCategories[1], allCategories[6]]);
+    var mousemove = CreateEvent('mousemove');
+    mousemove.pageX = 115;
+    mousemove.pageY = 574;
+    var mouseout = CreateEvent('mousemove');
+    var legends = $(container).find('.legend');
+    $(legends[0]).find('svg>rect')[0].dispatchEvent(mousemove);
+    $(legends[0]).find('svg>rect')[0].dispatchEvent(mouseout);
+    expect($(container).find('.color-legend-popup').text()).toBe('100 - 200');
+    $(legends[1]).find('svg>rect')[0].dispatchEvent(mousemove);
+    expect($(container).find('.color-legend-popup').text()).toBe('46');
+    $(legends[1]).find('svg>rect')[0].dispatchEvent(mouseout);
+    done();
+  });
+});
```

Comment:
Wow, how do you notice such small misses, thank you for such careful review.
I think I tested popup mouseout event, did the coverage says otherwise? Do you know how to see the line by line code coverage for the colorLegend file?

Reply:
https://codecov.io/gh/OpenGeoscience/geojs/src/2d_legend/src/ui/colorLegendWidget.js

Also, if you run the tests with coverage on locally, it will dump an html coverage report that you can browse.

Reply:
@jbeezley, Thank you! That's really helpful.